controller/controller.go

```diff
@@ -248,75 +246,85 @@ func (c *ObjectStorageController) runWorker(ctx context.Context) {
 
 func (c *ObjectStorageController) processNextItem(ctx context.Context) bool {
 	// Wait until there is a new item in the working queue
-	op, quit := c.queue.Get()
+	uuidInterface, quit := c.queue.Get()
 	if quit {
 		return false
 	}
 
+	uuid := uuidInterface.(types.UID)
+	var err error
 	// With the lock below in place, we can safely tell the queue that we are done
 	// processing this item. The lock will ensure that multiple items of the same
 	// name and kind do not get processed simultaneously
-	defer c.queue.Done(op)
+	defer c.queue.Done(uuid)
 
 	// Ensure that multiple operations on different versions of the same object
 	// do not happen in parallel
-	c.OpLock(op)
-	defer c.OpUnlock(op)
+	c.OpLock(uuid)
+	defer c.OpUnlock(uuid)
+
+	op := c.opMap[uuid]
 
-	var err error
 	switch o := op.(type) {
 	case addOp:
 		add := *o.AddFunc
+		objMeta := o.Object.(metav1.Object)
+		name := objMeta.GetName()
+		ns := objMeta.GetNamespace()
 		err = add(ctx, o.Object)
+		if err == nil {
+			o.Indexer.Add(o.Object)
+		} else {
+			glog.Errorf("Error adding %s %s: %v", ns, name, err)
+		}
 	case updateOp:
 		update := *o.UpdateFunc
+		objMeta := o.OldObject.(metav1.Object)
+		name := objMeta.GetName()
+		ns := objMeta.GetNamespace()
 		err = update(ctx, o.OldObject, o.NewObject)
+		if err == nil {
+			o.Indexer.Update(o.NewObject)
+		} else {
+			glog.Errorf("Error updating %s %s: %v", ns, name, err)
+		}
 	case deleteOp:
 		delete := *o.DeleteFunc
+		objMeta := o.Object.(metav1.Object)
+		name := objMeta.GetName()
+		ns := objMeta.GetNamespace()
 		err = delete(ctx, o.Object)
+		if err == nil {
+			o.Indexer.Delete(o.Object)
+		} else {
+			glog.Errorf("Error deleting %s %s: %v", ns, name, err)
+		}
 	default:
 		panic("unknown item in queue")
 	}
 
 	// Handle the error if something went wrong
-	c.handleErr(err, op)
+	c.handleErr(err, uuid)
 	return true
 }
 
-func (c *ObjectStorageController) OpLock(op interface{}) {
+func (c *ObjectStorageController) OpLock(op types.UID) {
 	c.GetOpLock(op).Lock()
 }
 
-func (c *ObjectStorageController) OpUnlock(op interface{}) {
+func (c *ObjectStorageController) OpUnlock(op types.UID) {
 	c.GetOpLock(op).Unlock()
 }
 
-func (c *ObjectStorageController) GetOpLock(op interface{}) *sync.Mutex {
-	var key string
-	var ext string
-
-	switch o := op.(type) {
-	case addOp:
-		key = o.Key
-		ext = fmt.Sprintf("%v", o.AddFunc)
-	case updateOp:
-		key = o.Key
-		ext = fmt.Sprintf("%v", o.UpdateFunc)
-	case deleteOp:
-		key = o.Key
-		ext = fmt.Sprintf("%v", o.DeleteFunc)
-	default:
-		panic("unknown item in queue")
-	}
+func (c *ObjectStorageController) GetOpLock(op types.UID) *sync.Mutex {
+	lockKey := op
+	c.lockerLock.Lock()
+	defer c.lockerLock.Unlock()
```

Comment:
should the `defer` be after L326 to handle the case of c.locker == nil?

Reply:
Even though the docs say that deferred funcs can be called anytime after the defer statement itself, my understanding is that it always gets called only after returning from the function it is called from. i.e. only after the defer starts going out of scope.